### PR TITLE
ENH: `CompositionSet`-based solver API

### PR DESCRIPTION
**Pull request by richardotis**
- Created new `SystemState` and `SystemSpecification` objects. `SystemSpecification` captures the desired result of a particular calculation and associated convergence criteria, while `SystemState` represents the state of a calculation. The pre-existing `Solver` object now uses a `SystemSpecification` to evolve a `SystemState` object until it achieves the convergence criteria in the specification (this is the specification's `run_loop` method). `SystemSpecification` provides a convenience method, `get_new_state`, which allows users to pass in a list of `CompositionSet` objects and get back a `SystemState`.
- Removed `update_composition_sets` and `solve_and_update` in eqsolver

`Solver.solve` can still be used as normal, but because the solver internals have been modularized, users have more control over the minimization process now. For example, solves can now be performed with a list `CompositionSet` objects as both input and output (mutated in place).
```python
# composition_sets is a list of CompositionSet representing a starting point
# conditions is a dict mapping state variable strings to floats
solver = Solver(remove_metastable=False) # will not delete CompositionSets found to be metastable
spec = solver.get_system_spec(composition_sets, conditions)
state = spec.get_new_state(composition_sets)
converged = spec.run_loop(state, 1000)
# composition_sets has been mutated with the result (solution meets convergence criteria if converged is True)
```

This is also useful for toggling the `fixed` state of a list of `CompositionSet` objects in place, which makes it convenient to do sequences of minimization operations with varying conditions.

### CODE REVIEW

File: pycalphad/core/minimizer.pyx [hunk 1]
```diff
@@ -586,6 +650,10 @@ cpdef solve_state(SystemSpecification spec, SystemState state):
         comp_idx = spec.fixed_chemical_potential_indices[chempot_idx]
         state.chemical_potentials[comp_idx] = spec.initial_chemical_potentials[comp_idx]
 
+    state.largest_chemical_potential_difference = -np.inf
+    for comp_idx in range(spec.num_components):
+        state.largest_chemical_potential_difference = max(state.largest_chemical_potential_difference, state.chemical_potentials[comp_idx] - state.previous_chemical_potentials[comp_idx])
```
Comment (bocklund) [inline pycalphad/core/minimizer.pyx:655]: Should it be absolute value?

```suggestion
        state.largest_chemical_potential_difference = max(state.largest_chemical_potential_difference, abs(state.chemical_potentials[comp_idx] - state.previous_chemical_potentials[comp_idx]))
```

File: pycalphad/core/minimizer.pyx [hunk 2]
```diff
@@ -549,12 +601,24 @@ cdef class SystemState:
...
+                self.metastable_phase_iterations[idx] += 1
+
 
 cpdef solve_state(SystemSpecification spec, SystemState state):
     cdef double[::1,:] equilibrium_matrix  # Fortran ordering required by call into lapack
     cdef double[::1] equilibrium_soln
     cdef int chempot_idx, comp_idx, num_stable_phases, num_fixed_phases, num_fixed_components, num_free_variables
 
+    if (state.mass_residual > 10) and (np.any(np.abs(state.chemical_potentials) > 1.0e10)):
+        state.chemical_potentials[:] = spec.initial_chemical_potentials
```
[the remark this answers is not in the document]
Comment (richardotis) [reply pycalphad/core/minimizer.pyx:619]: I've got Cython coverage working locally (`build_ext --inplace`) but not with the wheels. Anyway, the answer to your question is no, that condition is not triggered by the current test suite.

[the remark this answers is not in the document]
Comment (richardotis) [reply pycalphad/core/minimizer.pyx]: I removed the check. I don't think it's necessary anymore.

**Comment by bocklund** [comment]
Thanks for getting coverage working. Looks like Cython code isn't far from the rest of the repo in coverage. For the minimizer, it looks like a fixed compset + free state variable would cover many of the missed lines (not a blocker for this PR, IMO).

**Comment by bocklund** [comment]
Is this ready for re-review now?

**Comment by richardotis** [comment]
@bocklund It's ready as of `[4d3910d](https://github.com/pycalphad/pycalphad/pull/415/commits/4d3910d84c73200463d8076af1ef758dc8082147)`. I made a small refactor to `solve_state` which doesn't do much here, but it makes it easier to implement dot derivatives for some uncommitted local changes which are allowing me to compute the key differential quantity in Equation 78 of Sundman et al, 2015. I thought it made sense to get it in now as part of this review.

**Review by bocklund — APPROVED** [review]

**Pull request merged**